tensorrt_llm/_torch/modules/linear.py

```diff
@@ -628,6 +628,125 @@ def load_weights_fused_gate_up_linear(self, module: Linear,
         copy_weight(module.alpha, alpha)
 
 
+class W4A8MXFP4FP8LinearMethod(LinearMethodBase):
+
+    def create_weights(self, module: Linear, in_features: int,
+                       out_features: int, bias: bool, dtype: torch.dtype):
+        module.scaling_vector_size = 32
+        assert module.in_features % module.scaling_vector_size == 0, f"in_features {module.in_features} must be divisible by scaling_vector_size {module.scaling_vector_size}"
+        # Quantized weights
+        module.weight = Parameter(torch.empty(
+            [module.out_features, module.in_features // 2],
+            dtype=fp4_utils.float4_e2m1x2),
+                                  requires_grad=False)
+
+        # FP8 per-block scaling factors. dtype must be aligned with SF_DTYPE
+        # Padding is required. See computeSFSize in quantization.h
+        nrows = fp4_utils.pad_up(module.out_features, 128)
+        ncols = fp4_utils.pad_up(
+            module.in_features // module.scaling_vector_size, 4)
+        module.weight_scale = Parameter(torch.empty(
+            [nrows * ncols], dtype=fp4_utils.float4_sf_dtype),
+                                        requires_grad=False)
+
+        if bias:
+            module.bias = Parameter(torch.empty((out_features), dtype=dtype),
+                                    requires_grad=False)
+        else:
+            module.register_parameter("bias", None)
+
+    def apply(self, module: Linear, input: torch.Tensor,
+              bias: Optional[torch.Tensor]):
+        if input.dtype != torch.float8_e4m3fn:
+            # Dynamic quantization
+            fp8_input, input_scale = torch.ops.tensorrt_llm.quantize_e4m3_per_tensor(
+                input)
+            input_scale = input_scale.to(torch.float32)
+        else:
+            fp8_input, input_scale = input
```

Comment:
Why is `input_scale` read from input? Should it be read from the check point instead?

Reply:
If previous operation (attention kernel) fuses FP8 quantization then a tuple (fp8_input, input_scale) will be passed to linear forward.